utils/stickel.py

```diff
@@ -0,0 +1,160 @@
+            else:
+                self.smooth_y(lambda_param)
+            plt.plot(self.xdata, self.yhat, '.', label="Smoothed data", **kwargs)
+
+        if self.xlabel is not None:
+            plt.xlabel(self.xlabel)
+        if self.ylabel is not None:
+            plt.ylabel(self.ylabel)
+
+    def diff(self):
+        self.dx = np.diff(self.xdata)
+        self.dy = np.diff(self.ydata)
+
+        self.xmid = (self.xdata[:-1] + self.xdata[1:])/2
+        self.dydx = self.dy / self.dx
+
+    def diff_smooth(self, lambda_param=None):
+        if lambda_param is None:
+            if self.lambda_param is None:
+                raise ValueError("no lambda_param set, required for calculating smoothed derivative")
+        else:
+            self.smooth_y(lambda_param)
+
+        self.dyhat = np.diff(self.yhat)
+        self.dyhatdx = self.dyhat / self.dx
+
+    def plot_deriv(self, smooth=False, lambda_param=None, **kwargs):
+        if self.xmid is None or self.dydx is None:
+            self.diff()
+
+        plt.plot(self.xmid, self.dydx, '.', label="Orig numerical deriv.", **kwargs)
+
+        if smooth:
+            self.diff_smooth(lambda_param=lambda_param)
+            plt.plot(self.xmid, self.dyhatdx, '.', label="Smoothed deriv.", **kwargs)
+
+        if self.xlabel is not None:
+            plt.xlabel(self.xlabel)
+        if self.dylabel is not None:
+            plt.ylabel(self.dylabel)
+
+    def interp_smooth_deriv(self, **kwargs):
+        """kwargs applied to scipy.interpolate.interp1d.
+        """
+        if self.dyhatdx is None:
+            raise ValueError("Must create smoothed deriv (e.g., using diff_smooth()) before calling interp_smooth_deriv()")
+        self.dyhatdx_interp = interp1d(self.xmid, self.dyhatdx, kind='cubic', assume_sorted=True, **kwargs)
+
+    def unwrap(self, discont=np.pi):
+        self.ydata = np.unwrap(self.ydata, discont)
+
+    def smooth_y(self, lambda_param):
+        if lambda_param != self.lambda_param:
+            d = 3 # Using two orders higher than the derivative I want (1st), as recommended in Stickel (2010).
+            D = self._Dmatrix(d)
+            U = self._midpointtruncmatrix(d)
+            M = np.identity(self.N)
+            delta = np.trace(np.matmul(np.transpose(D), D)) / (self.N**(d+2))
+
+            if self.yerr is not None:
+                W = np.diag(1 / self.yerr**2)
+            else:
+                W = np.identity(self.N)
+
+            MW  = np.matmul(np.transpose(M), W)
+            MWM = np.matmul(MW, M)
+            MWy = np.matmul(MW, self.ydata)
+            DUD = np.matmul(np.matmul(np.transpose(D), U), D)
+            inv = np.linalg.pinv(MWM + lambda_param/delta*DUD)
+
+            self.yhat = np.matmul(inv, MWy)
+            self.lambda_param = lambda_param
+
+    def _midpointmatrix(self, d):
+        B1 = np.roll(self.xdata, 1)
+        B2 = np.roll(self.xdata, -1)
+        B1[0]  = self.xdata[0]
+        B2[-1] = self.xdata[-1]
+        
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Trailing whitespace](https://app.codacy.com/app/CIRA-Pulsars-and-Transients-Group/vcstools/pullRequest?prid=4645369)